SIGSEGV in kube.GetResourcesWithLabel

Noticed that an application controller had exited with the following error:
```
time="2018-08-21T20:02:20Z" level=info msg="cache miss: apires|https://CLUSTER-URL-REDACTED"
time="2018-08-21T20:02:20Z" level=warning msg="Failed to watch app resources on https://CLUSTER-URL-REDACTED: Get https://CLUSTER-URL-REDACTED/api: dial tcp: lookup CLUSTER-URL-REDACTED on 100.64.0.10:53: no such host, retrying in 10s"
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x45676c]

goroutine 7342065 [running]:
github.com/argoproj/argo-cd/util/kube.GetResourcesWithLabel.func1(0xc4213abe80, 0x1a57840, 0xc42106fcb0, 0x1572e47, 0x21, 0xc420dcb670, 0xe, 0xc4211174d0, 0xc420c53700)
        /root/go/src/github.com/argoproj/argo-cd/util/kube/kube.go:331 +0x3a3
created by github.com/argoproj/argo-cd/util/kube.GetResourcesWithLabel
        /root/go/src/github.com/argoproj/argo-cd/util/kube/kube.go:316 +0x695
```

--- a/util/kube/kube.go
+++ b/util/kube/kube.go
@@ -373,10 +373,11 @@ func DeleteResourceWithLabel(config *rest.Config, namespace string, labelName st
 		return err
 	}
 
-	var resourceInterfaces []struct {
+	type resClient struct {
 		dynamic.ResourceInterface
-		bool
+		deleteCollectionSupported bool
 	}
+	var resourceInterfaces []resClient
 
 	for _, apiResourcesList := range resources {
 		for i := range apiResourcesList.APIResources {
@@ -396,10 +397,10 @@ func DeleteResourceWithLabel(config *rest.Config, namespace string, labelName st
 			}
 
 			if deleteCollectionSupported || deleteSupported {
-				resourceInterfaces = append(resourceInterfaces, struct {
-					dynamic.ResourceInterface
-					bool
-				}{dclient.Resource(&apiResource, namespace), deleteCollectionSupported})
+				resourceInterfaces = append(resourceInterfaces, resClient{
+					dclient.Resource(&apiResource, namespace),
+					deleteCollectionSupported,
+				})
 			}
 		}
 	}
@@ -410,13 +411,10 @@ func DeleteResourceWithLabel(config *rest.Config, namespace string, labelName st
 	var wg sync.WaitGroup
 	wg.Add(len(resourceInterfaces))
 
-	for i := range resourceInterfaces {
-		client := resourceInterfaces[i].ResourceInterface
-		deleteCollectionSupported := resourceInterfaces[i].bool
-
-		go func() {
+	for _, c := range resourceInterfaces {
+		go func(client resClient) {
 			defer wg.Done()
-			if deleteCollectionSupported {
+			if client.deleteCollectionSupported {
 				err = client.DeleteCollection(&metav1.DeleteOptions{
 					PropagationPolicy: &propagationPolicy,
 				}, metav1.ListOptions{LabelSelector: fmt.Sprintf("%s=%s", labelName, labelValue)})
@@ -445,7 +443,7 @@ func DeleteResourceWithLabel(config *rest.Config, namespace string, labelName st
 					}
 				}
 			}
-		}()
+		}(c)
 	}
 	wg.Wait()
 	return asyncErr

--- a/util/kube/kube.go
+++ b/util/kube/kube.go
@@ -330,12 +330,11 @@ func GetResourcesWithLabel(config *rest.Config, namespace string, labelName stri
 
 	var asyncErr error
 	var result []*unstructured.Unstructured
-
 	var wg sync.WaitGroup
+	var lock sync.Mutex
 	wg.Add(len(resourceInterfaces))
-	for i := range resourceInterfaces {
-		client := resourceInterfaces[i]
-		go func() {
+	for _, c := range resourceInterfaces {
+		go func(client dynamic.ResourceInterface) {
 			defer wg.Done()
 			list, err := client.List(metav1.ListOptions{
 				LabelSelector: fmt.Sprintf("%s=%s", labelName, labelValue),
@@ -350,11 +349,13 @@ func GetResourcesWithLabel(config *rest.Config, namespace string, labelName stri
 				labels := item.GetLabels()
 				if labels != nil {
 					if value, ok := labels[labelName]; ok && value == labelValue {
+						lock.Lock()
 						result = append(result, &item)
+						lock.Unlock()
 					}
 				}
 			}
-		}()
+		}(c)
 	}
 	wg.Wait()
 	return result, asyncErr

